### DISCUSSION TITLE
Expression to match domain and all subdomains.

### DISCUSSION THREAD

**Discussion by WhyNotHugo** [Cookie-AutoDelete/Cookie-AutoDelete#1433] [General] [answered]
I expected the expression `example.com` to match `app.example.com`, but it does not. The wiki suggests that [subdomain prefixes are stripped](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/wiki/Documentation#enter-expression), but this doesn't seem to be the case. I guess `example.com` applies only if the domain matches the regex `^example.com$`?

I guess I could add `example.com` **and** `*.example.com`, but I'd prefer the simplicity of adding just ONE rule for this case that repeats over and over.

Shouldn't `example.com` just match that domain and _any_ subdomain? I'm sure some other syntax could be used for cases where only the apex is desired (e.g.: `^example.com`?).

**Comment by WhyNotHugo** [comment]
Oh, it seems that `*example.com` works. Shall I update the wiki to reflect this?

**Comment by kennethtran93** [comment, marked as answer]
The only exception was `www.domain.com` matching domain. All other subdomains need to be added, or use `*.domain.com`
 (this will also match the main domain.com as well). There is already an entry for this, so not sure what part of the wiki needs to be adjusted. 

Be cautious of using `*domain.com`, because somebody could have a domain with `baddomain.com` and it would match.

**Reply by WhyNotHugo** [reply]
> Be cautious of using *domain.com, because somebody could have a domain with baddomain.com and it would match.

Good catch, thanks for mentioning it.

> use *.domain.com

I also want to match `domain.com`. E.g.: the apex.

**Reply by WhyNotHugo** [reply]
(but I'd rather do this in one single rule; this is a VERY common scenario)

**Reply by kennethtran93** [reply]
Did I not mention that using `*.example.com` would also cover example.com?

I believe it is also mentioned in the wiki as well.

**Reply by kennethtran93** [reply]
![Screenshot_20220729-112124_Firefox.png](https://user-images.githubusercontent.com/6724477/181821187-50e8e600-8bca-40e1-8072-c4c1d4fba3d9.png)

From the wiki

**Reply by WhyNotHugo** [reply]
Oh, so it's not a regular glob expression; it's a slightly more specific syntax. So the `*.` imply "this domain and others", but a dot by itself is to prevent the `prefix-example.com` attacks, right?

**Reply by kennethtran93** [reply]
Yes. All dots in this case is specifically a period, not 'any character', unless you actually write your own regular expression (start and end with the slash).

**Reply by WhyNotHugo** [reply]
> All dots in this case is specifically a period

Except a leading `*.`, in which case this dot isn't treated as a period, but this pair of characters becomes a special expression that matches either of these:

- Any subdomain (e.g.: any string ending with a `.`, or, in other words, the glob `*.`).
- Empty string (e.g.: _no_ period!).


However, this two-character expression DOES NOT match anything that doesn't end with a period.

I think this covers all corner cases, right?

**Reply by kennethtran93** [reply]
> Except a leading `*.`, in which case this dot isn't treated as a period, but this pair of characters becomes a special expression that matches either of these:
> 
>     * Any subdomain (e.g.: any string ending with a `.`, or, in other words, the glob `*.`).
> 
>     * Empty string (e.g.: _no_ period!).

why would there be an empty string anyway?


> However, this two-character expression DOES NOT match anything that doesn't end with a period.

Not sure what you're getting at here, considering this only applies at the beginning of the expression.  If it is found anywhere else, it will be interpreted as is (i.e. match 0 or more of the preceding character, followed by any character except line breaks) as per regular expressions.

To clarify, we only check if it starts with `*.`, not anywhere else in the string.  If so, we translate it to `(^|\.)`, and if there's any regular expression characters used but not stated inside a regular expression, we automatically prepend the escape `\` character to it so the regular expression tester will parse it as is.  If `*` wildcard character is used, we automatically convert that to `.*` for regular expression usage.  If it isn't a regular expression, we will make it one, and append a `$` in the end to ensure the actual domain is captured.

So `*.google.com` would become `/(^|\.)google\.com$/`, to which `sub.domain.google.com`, `sub.google.com`, and `google.com` would match.  Essentially means it would match as long as the main domain ends with `google.com`.

Likewise, `google.com` would become `/^google\.com$/`.
